src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,518 @@
+            SimulationValues[qubit.Id];
+
+        // SetValue sets the simulation value at the corresponding index.
+        private void SetValue(Qubit qubit, bool value)
+        {
+            SimulationValues[qubit.Id] = value;
+        }
+
+        // InvertValue inverts the simulation value at the corresponding index.
+        private void InvertValue(Qubit qubit)
+        {
+            SimulationValues[qubit.Id] = !SimulationValues[qubit.Id];
+        }
+
+        // When allocating a qubit whose index equals or exceeds the current
+        // number of simulation values stored in the bit array, the bit array is
+        // resized to twice its current size.
+        public override void OnAllocateQubits(IQArray<Qubit> qubits)
+        {
+            if (qubits.Count == 0) return;
+
+            var maxId = qubits.Max(q => q.Id);
+            while (maxId >= SimulationValues.Length)
+            {
+                SimulationValues.Length *= 2;
+            }
+        }
+
+        // When releasing qubits that are not in the zero state, their value is
+        // reset to false.  If the ThrowOnReleasingQubitsNotInZeroState property
+        // is assigned true, an exception is thrown.
+        public override void OnReleaseQubits(IQArray<Qubit> qubits)
+        {
+            foreach (var qubit in qubits)
+            {
+                if (GetValue(qubit))
+                {
+                    if (ThrowOnReleasingQubitsNotInZeroState)
+                    {
+                        throw new ReleasedQubitsAreNotInZeroState();
+                    }
+                    SetValue(qubit, false);
+                }
+            }
+        }
+
+        // The borrowed qubits' states are stored on a stack together with their current value.
+        // The value is used for comparison when the qubit is returned.
+        public override void OnBorrowQubits(IQArray<Qubit> qubits, long allocatedForBorrowingCount)
+        {
+            foreach (var qubit in qubits)
+            {
+                Stack<bool> stack;
+                if (!borrowedQubitStates.TryGetValue(qubit, out stack))
+                {
+                    stack = new Stack<bool>();
+                    borrowedQubitStates.Add(qubit, stack);
+                }
+                stack.Push(GetValue(qubit));
+            }
+        }
+
+        // A qubit must be returned in its original value when borrowed.
+        public override void OnReturnQubits(IQArray<Qubit> qubits, long releasedOnReturnCount)
+        {
+            foreach (var qubit in qubits)
+            {
+                var stack = borrowedQubitStates[qubit];
+                if (stack.Pop() != GetValue(qubit))
+                {
+                    throw new ExecutionFailException("Borrowed qubit returned not in the original state");
+                }
+
+                if (stack.Count == 0)
+                {
+                    borrowedQubitStates.Remove(qubit);
+                }
```

Comment:
Does it make sense to remove empty stacks, or is it better to keep them around once created to avoid churn in memory?